### DISCUSSION TITLE
Update a color value on drag end

### DISCUSSION THREAD

**Discussion by marcelpi** [pmndrs/leva#325] [Q&A] [answered]
Hi there,

I'm using ThreeJS and Leva to update the color of geometry and whenever I change the color from the colorpicker all the items adjust.

The problem is I'm having over 10k cubes to update and on each change the whole computation starts over which results in laggy Leva interface. Basically the slider stops until the computation is finished.

Is there a way to update the color value only After the dragging has ended? Or detect the end of drag and get the color value in a separate variable?

Please let me know, thanks.

**Comment by gsimone** [comment, marked as answer]
you can use the onChange property to apply the change in a debounced way

```js
const debouncedApply = debounce((color) => { applyColor(color) }, 1000)

useControls({
   color: 0xff0000,
   onChange: newColor => {
		debouncedApply(newColor)
   }
})
```

Alternatively use onEditEnd instead:

```js
useControls({
   color: 0xff0000,
   onEditEnd: newColor => {
		applyColor(newColor)
   }
})
```

**Reply by dbismut** [reply]
There's also `onEditEnd` that might come useful.

**Reply by marcelpi** [reply]
> you can use the onChange property to apply the change in a debounced way
> 
> ```js
> const debouncedApply = debounce((color) => { applyColor(color) }, 1000)
> 
> useControls({
>    color: 0xff0000,
>    onChange: newColor => {
> 		debouncedApply(newColor)
>    }
> })
> ```
> 
> Of course implementation details are left to you.

thanks @gsimone. and what about the debounce function? is that a setTimeout function I need to define myself?

**Reply by dbismut** [reply]
Yes. Or you could use `onEditEnd` that only fires when you release the slider as mentioned above.

**Reply by marcelpi** [reply]
You're right, will use onEditEnd which makes perfect sense.
Thanks for your feedback guys.

**Reply by dbismut** [reply]
@gsimone I'm giving the answer karma to you 😇
